src/track/beatgrid.cpp

```diff
@@ -3,28 +3,54 @@
 
 #include "track/beatgrid.h"
 #include "util/math.h"
+#include "maths/MathUtilities.h"
 
 static const int kFrameSize = 2;
 
 struct BeatGridData {
     double bpm;
     double firstBeat;
+    double firstPhrase;
 };
 
+
 class BeatGridIterator : public BeatIterator {
   public:
-    BeatGridIterator(double dBeatLength, double dFirstBeat, double dEndSample)
+    BeatGridIterator(double dBeatLength, double dFirstBeat, double dEndSample,
+    		double dFirstPhraseBegin, int beatsPerBar, int barsPerPhrase)
             : m_dBeatLength(dBeatLength),
               m_dCurrentSample(dFirstBeat),
-              m_dEndSample(dEndSample) {
+              m_dEndSample(dEndSample),
+			  m_dPhraseBeginSample(dFirstPhraseBegin),
+              m_beatsPerBar(beatsPerBar),
+              m_barsPerPhrase(barsPerPhrase),
+              m_beatCounter(0) {
     }
 
+
     virtual bool hasNext() const {
         return m_dBeatLength > 0 && m_dCurrentSample <= m_dEndSample;
     }
 
-    virtual double next() {
-        double beat = m_dCurrentSample;
+    virtual BeatData next() {
+        BeatData beat;
+        beat.sample = m_dCurrentSample;
+
+
+        if (m_dPhraseBeginSample > m_dCurrentSample or m_dCurrentSample > m_dEndSample) {
```

Comment:
Please avoid to use `or`, that's very uncommon in C++

Reply:
use || instead